compiler/include/metadata.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2020-2023 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef _ATTRIBUTE_H_
+#define _ATTRIBUTE_H_
+
+#include <memory>
+#include <cstdint>
+#include <vector>
+
+enum LLVMMetadataType {
+  LAT_NO_VALUE, LAT_INT, LAT_BOOL, LAT_STRING, LAT_ATTRIBUTE
+};
+struct LLVMMetadata;
+using LLVMMetadataPtr = std::shared_ptr<LLVMMetadata>;
+using LLVMMetadataList = std::vector<LLVMMetadataPtr>;
+struct LLVMMetadata {
+  const char* key;
+  // TODO: This absolutely should be a C++17 variant
```

Comment:
I think we do require C++17 don't we? Do you want an additional protection here?

Reply:
We sadly only use C++17 if using LLVM 16

Reply:
And mostly my strong preference here for variant is because it will catch more programmer mistakes at compile time, give nicer errors if we make mistakes at runtime, and removes the need to explicitly construct/destruct the union with shared_ptr.